### ISSUE TITLE
[BUG Gaiax] Engine terminated unexpectedly

### ISSUE THREAD

**Issue by longbinlai** [alibaba/GraphScope#978]
**Describe the bug**
The engine terminated unexpectedly within a subtask. The codes are like:
```
    conf.set_workers(4);
    let mut result = pegasus::run(conf, move || {
        let index = pegasus::get_current_worker().index;
        move |input, output| {
            let src = if index == 0 {
                input.input_from(
                    MAP3.get(&Tuple(src_v, 1)).unwrap()
                        .iter()
                        .map(|t| (t.0, t.1))
                )?
            } else {
                input.input_from(vec![].into_iter())?
            };
            src.flat_map(|v| Ok(
                MAP3.get(&v.into()).unwrap().iter().map(|t| (t.0, t.1)))
            )?
                .limit(expected_cnt as u32)?
                // .map(|(v, _)| Ok(v))?
                .apply(|sub| {
                    sub.repartition(|v| Ok(v.0))
                        .flat_map(|v| {
                            Ok(MAP3.get(&v.into()).unwrap().iter().map(|t| (t.0, t.1)))
                        })?
                        .repartition(|(v, _)| Ok(*v))
                        .map(|x| Ok(x))?
                        .filter(|t| Ok(t.1 == 3))?
                        .repartition(|v| Ok(v.0))
                        .flat_map(|v| {
                            Ok(MAP4.get(&v.into()).unwrap().iter().map(|t| (t.0, t.1)))
                        })?
                        .limit(1)?
                        .count()
                })?
                .filter_map(|(v, cnt)| if cnt == 0 { Ok(None) } else { Ok(Some(v)) })?
                .count()?
                .sink_into(output)
        }
    })
        .expect("build job failure");
```

**To Reproduce**
1. Go to the branch 'gaia-x-bugs', cd research/gaia-x/pegasus/pegasus
2. Run the test case AS: cargo test -- apply_flatmap_limit_unexpected_results4

**Expected behavior**
The execution should terminate neatly. 

**Screenshots**
![Screen Shot 2021-11-12 at 11 39 36](https://user-images.githubusercontent.com/7007589/141405077-8e46891b-16f2-4a93-92e6-48bdb0a5adb3.png)

**Additional context**
A related bug fix is with the commit: https://github.com/alibaba/GraphScope/commit/b85fae22fe524bd86163fb932f335b9c48695734

But when I add the following fragment, new bug shows. 
Note that this is corresponding to a Gremlin bug that we found previously.
```
.map(|x| Ok(x))?
.filter(|t| Ok(t.1 == 3))?
.repartition(|v| Ok(v.0))
.flat_map(|v| {
      Ok(MAP4.get(&v.into()).unwrap().iter().map(|t| (t.0, t.1)))
})?
```

**Comment by longbinlai** [comment]
Bug fixed.

**Comment by longbinlai** [comment]
The bug has been fixed with the commit: c50046b
